flink-runtime/src/main/java/org/apache/flink/runtime/shuffle/ShuffleEnvironment.java

```diff
@@ -0,0 +1,163 @@
+ *
+ * <h3>Input gate management.</h3>
+ *
+ * <p>The interface implements a factory for the task input gates: {@code createInputGates}.
+ * The created gates are grouped per task and handed over to the task thread upon its startup.
+ * The task is responsible for the gates' lifecycle from that moment.
+ *
+ * <p>When tha task is deployed and the input gates are created, it can happen that not all consumed partitions
+ * are known at that moment e.g. because their producers have not been started yet.
+ * Therefore, the {@link ShuffleEnvironment} provides a method {@code updatePartitionInfo} to update them
+ * externally, ouside of the task thread, when the producer becomes known.
+ */
+public interface ShuffleEnvironment {
+
+	/**
+	 * Start the internal related services upon {@link TaskExecutor}'s startup.
+	 *
+	 * @return a port to connect to the task executor for shuffle data exchange, -1 if only local connection is possible.
+	 */
+	int start() throws IOException;
+
+	/**
+	 * Shutdown the internal related services upon {@link TaskExecutor}'s stop.
+	 */
+	void shutdown();
+
+	/**
+	 * Factory method for the task's {@link ResultPartitionWriter}s.
+	 *
+	 * @param taskName the task name, used for logs
+	 * @param executionAttemptID execution attempt id of the task
+	 * @param resultPartitionDeploymentDescriptors descriptors of the partition, produced by the task
+	 * @param outputGroup shuffle specific group for output metrics
+	 * @param buffersGroup shuffle specific group for buffer metrics
+	 * @return array of the task's {@link ResultPartitionWriter}s
+	 */
+	ResultPartitionWriter[] createResultPartitionWriters(
+		String taskName,
+		ExecutionAttemptID executionAttemptID,
+		Collection<ResultPartitionDeploymentDescriptor> resultPartitionDeploymentDescriptors,
+		MetricGroup outputGroup,
+		MetricGroup buffersGroup);
+
+	/**
+	 * Release local resources occupied with the given partitions.
+	 *
+	 * @param partitionIds partition ids to release
+	 */
+	void releasePartitions(Collection<ResultPartitionID> partitionIds);
+
+	/**
+	 * Report unreleased partitions.
+	 *
+	 * @return collection of partitions which still occupy some resources locally on this task executor
+	 * and have been not released yet.
+	 */
+	Collection<ResultPartitionID> getUnreleasedPartitions();
+
+	/**
+	 * Factory method for the task's {@link InputGate}s.
+	 *
+	 * @param taskName the task name, used for logs
+	 * @param executionAttemptID execution attempt id of the task
+	 * @param partitionProducerStateProvider producer state provider to query whether the producer is ready for consumption
+	 * @param inputGateDeploymentDescriptors descriptors of the input gates, consumed by the task
+	 * @param parentGroup parent of shuffle specific metric group
+	 * @param inputGroup shuffle specific group for input metrics
+	 * @param buffersGroup shuffle specific group for buffer metrics
+	 * @return array of the task's {@link InputGate}s
+	 */
+	InputGate[] createInputGates(
```

Comment:
Why do we return an array here instead of a `Collection<InputGate>`? Is it because of the order of the input gates? Maybe we could state that the returned collection must have the same order as the `inputGateDeploymentDescriptors` collection.

Reply:
true, will change to collection and add the comment about the order